### PR TITLE
New Feature - Camo View Phase 2

### PR DESCRIPTION
**Pull request by sol3uk**
# Work In Progress - Camo View #3 

- Added new camo components
- Render camouflage view on another route
- Shared progress between different views
- Shared filtering between both views
- Fix some bad data whilst combing through and testing
- Rename solidColours to solidColors (to keep the Americans happy)

There's definitely some optimisation and best practises that could be done here, but I'm putting this up because I'm mostly happy with where it's at. 
Some things I need feedback on or would like to improve:
- I want to add a "Completion/Mastery" category, but it would require additional thought on how to display it, i.e. should we have gold, plat, poly and show the progress shown against it in some way? (make them non-toggle-able and show as read-only?)
- Filtering section doesn't make 100% sense when in camo view? Might need to add some clarity to it or change it around.
- Not sure if keeping the challenge only visible on hover is best? Could be nice to show it at the top level, don't want to crowd the screen though. Suggestions?
- Don't know if mobile optimisation is the best it could be in terms of card & image size
- Should we be making it more obvious when the camos/weapons are being affected by a filter?
- Can you think of a better way to do filtering without having 2 separate progress state/stores?

### CODE REVIEW

File: src/components/CamoComponent.vue [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+			return new URL(`../assets/camouflages/${convertToKebabCase(camoName)}.png`, import.meta.url)
+		},
+
+		imageExists(camoName) {
+			const url = this.imageUrl(camoName)
+			const img = new Image()
+			img.src = url
+			return img.height !== 0
+		},
```
Comment (carlssonemil) [inline src/components/CamoComponent.vue]: I don't think this is used anywhere and could be removed.
```suggestion

```

File: src/components/CamoComponent.vue [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+			if (camoName === 'Gold') {
+				return new URL('/gold-gradient.svg', import.meta.url)
+			} else if (camoName === 'Platinum') {
+				return new URL('/platinum-gradient.svg', import.meta.url)
+			} else if (camoName === 'Polyatomic') {
+				return new URL('/polyatomic-gradient.svg', import.meta.url)
+			}
+
+			return new URL(`../assets/camouflages/${convertToKebabCase(camoName)}.png`, import.meta.url)
+		},
```
Comment (carlssonemil) [inline src/components/CamoComponent.vue]: Can we change so this parameter is just named `camouflage`?
```suggestion
		imageUrl(camouflage) {
			if (camouflage === 'Gold') {
				return new URL('/gold-gradient.svg', import.meta.url)
			} else if (camouflage === 'Platinum') {
				return new URL('/platinum-gradient.svg', import.meta.url)
			} else if (camouflage === 'Polyatomic') {
				return new URL('/polyatomic-gradient.svg', import.meta.url)
			}

			return new URL(`../assets/camouflages/${convertToKebabCase(camouflage)}.png`, import.meta.url)
		},
```

File: src/components/CamosComponent.vue [hunk 3]
```diff
@@ -0,0 +1,143 @@
+<template>
+	<div>
+		<transition-group name="fade" tag="div" class="container">
+			<div
+				v-for="(category, title, index) in camos"
+				:key="title"
+				:data-index="index"
+				class="category">
+				<!-- TODO: Add double-click functionality -->
```
Comment (carlssonemil) [inline src/components/CamosComponent.vue]: This comment is probably not needed 😄 
```suggestion
```

File: src/components/layout/NavigationComponent.vue [hunk 4]
```diff
@@ -5,7 +5,8 @@
 			<p><span>Orion</span></p>
 		</router-link>
 		<div>
-			<router-link to="/">Camouflages</router-link>
+			<router-link to="/">Weapons</router-link>
+			<router-link to="/camo">Camouflages</router-link>
```
Comment (carlssonemil) [inline src/components/layout/NavigationComponent.vue]:
```suggestion
			<router-link to="/camouflages">Camouflages</router-link>
```

File: src/components/CamoComponent.vue [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+		}
+	},
+	props: {
+		camo: {
+			type: Object,
+			required: true,
+		},
+	},
+	computed: {
+		...mapState(useStore, ['camoRequirements']),
```
Comment (carlssonemil) [inline src/components/CamoComponent.vue]:
```suggestion
		...mapState(useStore, ['camouflageRequirements']),
```

File: src/components/CamoComponent.vue [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+			img.src = url
+			return img.height !== 0
+		},
+
+		requirementTooltip(camo) {
+			const requirement = this.camoRequirements[camo.category][camo.name]
+			return `Get the ${requirement.weapon} to level ${requirement.level} - ${
+				requirement.challenge || 'TBA'
+			}`
+		},
```
Comment (carlssonemil) [inline src/components/CamoComponent.vue]:
```suggestion
		requirementTooltip(camouflage) {
			const requirement = this.camouflageRequirements[camouflage.category][camouflage.name]
			return `Get the ${requirement.weapon} to level ${requirement.level} - ${
				requirement.challenge || 'TBA'
			}`
		},
```

File: src/components/CamoComponent.vue [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+
+		handleToggleCompleted(camo) {
+			const weaponName = this.camoRequirements[camo.category][camo.name].weapon
+			this.toggleCamouflageCompleted(weaponName, camo.name, this.isCompleted)
+			this.toggleIsCompleted()
+		},
```
Comment (carlssonemil) [inline src/components/CamoComponent.vue]:
```suggestion
		handleToggleCompleted(camouflage) {
			const weaponName = this.camouflageRequirements[camouflage.category][camouflage.name].weapon
			this.toggleCamouflageCompleted(weaponName, camouflage.name, this.isCompleted)
			this.toggleIsCompleted()
		},
		
```

File: src/components/CamoComponent.vue [hunk 8]
```diff
@@ -0,0 +1,211 @@
+<template>
+	<div class="camo">
+		<div
+			:key="camo.name"
```
Comment (carlssonemil) [inline src/components/CamoComponent.vue]: Probably don't need a `:key` here.
```suggestion
```

File: src/components/CamoComponent.vue [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+			:class="['camouflage']"
+			@click="handleToggleCompleted(camo)"
+			:content="requirementTooltip(camo)"
```
Comment (carlssonemil) [inline src/components/CamoComponent.vue]:
```suggestion
			@click="handleToggleCompleted(camouflage)"
			:content="requirementTooltip(camouflage)"
```

File: src/components/CamoComponent.vue [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+			v-tippy="{ placement: 'bottom' }">
+			<div :class="['inner', this.isCompleted ? 'completed' : '']">
+				<img
+				:src="imageUrl(camo.name)"
+				:alt="camo.name"
+				onerror="javascript:this.src='/base-gradient.svg'" />
+				<IconComponent class="complete" name="check" fill="#10ac84" />
+				<IconComponent class="remove" name="times" fill="#ee5253" />
+				<span>
+					{{ camo.name }}
```
Comment (carlssonemil) [inline src/components/CamoComponent.vue]:
```suggestion
					{{ camouflage.name }}
```

File: src/components/CamoComponent.vue [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+import { mapActions, mapState } from 'pinia'
+
+export default {
+	data() {
+		return {
+			isCompleted: this.camo.isCompleted,
+		}
+	},
+	props: {
+		camo: {
```
Comment (carlssonemil) [inline src/components/CamoComponent.vue]:
```suggestion
		camouflage: {
```

File: src/components/CamoComponent.vue [hunk 12]
```diff
@@ -0,0 +1,211 @@
+<template>
+	<div class="camo">
```
Comment (carlssonemil) [inline src/components/CamoComponent.vue]: Is this "camo" div really needed?

File: src/components/CamoComponent.vue [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+		<div
+			:key="camo.name"
+			:class="['camouflage']"
```
Comment (carlssonemil) [inline src/components/CamoComponent.vue]:
```suggestion
			class="camouflage"
```

File: src/components/CamosComponent.vue [hunk 14]
```diff
@@ -0,0 +1,143 @@
+<template>
+	<div>
+		<transition-group name="fade" tag="div" class="container">
+			<div
+				v-for="(category, title, index) in camos"
```
Comment (carlssonemil) [inline src/components/CamosComponent.vue]:
```suggestion
				v-for="(category, title, index) in camouflages"
```

File: src/components/CamosComponent.vue [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+						{{ title }}
+					</span>
+					<span v-tippy content="Camouflage progress for category">
+						{{ categoryProgress(title) }}
+					</span>
+				</h2>
+
+				<transition-group name="fade" tag="div" class="camos">
+					<CamoComponent v-for="camo in category" :key="camo.name" :camo="camo" />
+				</transition-group>
```
Comment (carlssonemil) [inline src/components/CamosComponent.vue]:
```suggestion
				<transition-group name="fade" tag="div" class="camouflages">
					<CamouflageComponent v-for="camouflage in category" :key="camouflage.name" :camouflage="camouflage" />
				</transition-group>
```

File: src/components/CamosComponent.vue [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+			</div>
+		</transition-group>
+
+		<div v-if="Object.keys(camos).length === 0" class="finished-placeholder">
```
Comment (carlssonemil) [inline src/components/CamosComponent.vue]:
```suggestion
		<div v-if="Object.keys(camouflages).length === 0" class="finished-placeholder">
```

File: src/components/CamosComponent.vue [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+			<p>You have completed all camouflage challenges 👏</p>
+		</div>
+	</div>
+</template>
+
+<script>
+import { mapState } from 'pinia'
+import { useStore } from '@/stores/store'
+// import allCamos from '../data/camos'
```
Comment (carlssonemil) [inline src/components/CamosComponent.vue]: Possible redundant comment? Unless you want to add it later? 🤔 

```suggestion
```

File: src/components/CamosComponent.vue [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+import CamoComponent from '@/components/CamoComponent.vue'
```
Comment (carlssonemil) [inline src/components/CamosComponent.vue]:
```suggestion
import CamouflageComponent from '@/components/CamouflageComponent.vue'
```

File: src/components/CamosComponent.vue [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+
+export default {
+	components: {
+		CamoComponent,
```
Comment (carlssonemil) [inline src/components/CamosComponent.vue]:
```suggestion
		CamouflageComponent,
```

File: src/components/CamosComponent.vue [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+	},
+
+	props: {
+		camos: {
```
Comment (carlssonemil) [inline src/components/CamosComponent.vue]:
```suggestion
		camouflages: {
```

File: src/components/CamosComponent.vue [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+	},
+
+	computed: {
+		...mapState(useStore, ['weapons']),
+	},
+
+	methods: {
+		categoryProgress(categoryTitle) {
+			let completed = 0
+			const categoryCamos = this.camos[categoryTitle].map((catCamo) => catCamo.name)
```
Comment (carlssonemil) [inline src/components/CamosComponent.vue]:
```suggestion
			const categoryCamos = this.camouflages[categoryTitle].map((catCamo) => catCamo.name)
```

File: src/data/camos/classic.js [hunk 22]
```diff
@@ -0,0 +1,11 @@
+import classic from '../requirements/camo/classic'
+const base = Object.keys(classic)
+
+const dlc = []
```
Comment (carlssonemil) [inline src/data/camos/classic.js:4]: Will there be DLC camouflages? 🤔

File: src/router/index.js [hunk 23]
```diff
@@ -1,6 +1,7 @@
 import Vue from 'vue'
 import VueRouter from 'vue-router'
 import HomeView from '../views/HomeView.vue'
+import CamoView from '../views/CamoView.vue'
```
Comment (carlssonemil) [inline src/router/index.js]: Let's import this dynamically instead, see comment below.
```suggestion
```

File: src/router/index.js [hunk 24]
```diff
@@ -20,6 +21,11 @@ const router = new VueRouter({
 			name: '404',
 			component: NotFoundView,
 		},
+		{
+			path: '/camo',
+			name: 'camo',
+			component: CamoView,
+		},
```
Comment (carlssonemil) [inline src/router/index.js]:
```suggestion
		{
			path: '/camouflages',
			name: 'camouflages',
			component: () => import('../views/CamouflagesView.vue'),
		},
```

File: src/stores/store.js [hunk 25]
```diff
@@ -3,15 +3,17 @@ import { defineStore } from 'pinia'
 import { filterObject } from '../utils/utils'
 import defaultWeapons from '../data/weapons'
 import defaultFilters from '../data/defaults/filters'
-import requirements from '../data/requirements'
+import weaponRequirements from '../data/weaponRequirements'
+import camoRequirements from '../data/camoRequirements'
```
Comment (carlssonemil) [inline src/stores/store.js]:
```suggestion
import camouflageRequirements from '../data/camouflageRequirements'
```

File: src/stores/store.js [hunk 26]
```diff
@@ -3,15 +3,17 @@ import { defineStore } from 'pinia'
...
 
 const token = import.meta.env.MODE === 'production' ? 'orion' : 'orion-dev'
 
 export const useStore = defineStore({
 	id: 'store',
 
 	state: () => ({
-		requirements: { ...requirements },
+		weaponRequirements: { ...weaponRequirements },
+		camoRequirements: { ...camoRequirements },
```
Comment (carlssonemil) [inline src/stores/store.js]:
```suggestion
		camouflageRequirements: { ...camouflageRequirements },
```

File: src/views/CamoView.vue [hunk 27]
```diff
@@ -0,0 +1,123 @@
...
+
+h1 {
+	margin-top: 75px;
+}
+
+h2 {
+	margin: 30px auto 0;
+	max-width: 450px;
+}
+</style>
```
Comment (carlssonemil) [inline src/views/CamoView.vue]: Let's remove the filtering for now. I'll make the `FilterComponent` more dynamic so it fits better 😄 The alert is probably not needed here either.
```suggestion
<template>
	<div class="container">
		<CamouflagesComponent :camouflages="camouflages" />
		<ProgressComponent />
	</div>
</template>

<script>
import { mapState } from 'pinia'
import { useStore } from '@/stores/store'
import { groupBy, filterObject } from '@/utils/utils'
import allCamos from '../data/camos'

import CamouflagesComponent from '@/components/CamouflagesComponent.vue'
import ProgressComponent from '@/components/ProgressComponent.vue'

export default {
	components: {
		CamosComponent,
		ProgressComponent,
	},

	computed: {
		...mapState(useStore, ['weapons']),
		
		camouflages() {
		  return groupBy(allCamos, (camouflage) => camouflage.category)
		},
	},
}
</script>

<style lang="scss" scoped>
.container {
	text-align: center;
}

h1 {
	margin-top: 75px;
}

h2 {
	margin: 30px auto 0;
	max-width: 450px;
}
</style>
```

**Review by carlssonemil — CHANGES_REQUESTED** [review]
I've looked it over and added some suggestions! I really like what you've done so far though! Great job! 🥳 ❤️ 

I hope I don't come across as too picky 🙈 

Also, a general question, to not make the review messier than it already is:
Do we really need to set `isCompleted` on the camouflage object? Can't we just find it on the weapon progress? I don't see the need to track it twice 🤔 

---

Regarding your questions:

> I want to add a "Completion/Mastery" category, but it would require additional thought on how to display it, i.e. should we have gold, plat, poly and show the progress shown against it in some way? (make them non-toggle-able and show as read-only?)

Hm.. 🤔 Is there really a need to have it be shown on the camouflage page though? Since it's connected to every weapon?

> Filtering section doesn't make 100% sense when in camo view? Might need to add some clarity to it or change it around.

I removed it in the review suggestion. I'll look into making it more dynamic so it fits into this view better!

> Not sure if keeping the challenge only visible on hover is best? Could be nice to show it at the top level, don't want to crowd the screen though. Suggestions?

I've thought about this before, in previous trackers, and after trying out different solutions I ended up deciding it looks the cleanest having it as a tooltip. Despite it being kinda hard to access on mobile, but I haven't had any complaints yet so it should most likely be fine as is! 😄 

> Don't know if mobile optimisation is the best it could be in terms of card & image size

I think it looks fine as is! We can adjust it later if needs be 😃 

> Should we be making it more obvious when the camos/weapons are being affected by a filter?

That's a good point! I'll look into adding some sort of visualization element that the weapons/camouflages are filtered.

> Can you think of a better way to do filtering without having 2 separate progress state/stores?

Not really sure what you mean here 🤔

File: src/components/CamoComponent.vue (hunk 12 above)
Comment (sol3uk) [reply src/components/CamoComponent.vue]: tidied this up, there was some unecessary css as well

File: src/data/camos/classic.js (hunk 22 above)
Comment (sol3uk) [reply src/data/camos/classic.js:4]: I would hope so, if they're doing unique camos per weapon it only makes sense they introduce more camos with the new seasons! 😄

File: src/views/CamoView.vue (hunk 27 above)
Comment (sol3uk) [reply src/views/CamoView.vue]: I've cleaned up the filtering and just left it with the initialisation code for getting the camouflage entities and progress. Yeah we can probably refactor the filtering code to be shared, I wasn't sure on best practices for that in vue and how they handle shared methods when working with active state.

**Comment by sol3uk** [comment]
> Also, a general question, to not make the review messier than it already is: Do we really need to set `isCompleted` on the camouflage object? Can't we just find it on the weapon progress? I don't see the need to track it twice 🤔

The main reason for needing this is just to have a base property available to map against. I could add it in programmatically or re-build a list of objects in memory, but that'd be more expensive

> > I want to add a "Completion/Mastery" category, but it would require additional thought on how to display it, i.e. should we have gold, plat, poly and show the progress shown against it in some way? (make them non-toggle-able and show as read-only?)

> Hm.. 🤔 Is there really a need to have it be shown on the camouflage page though? Since it's connected to every weapon?

I think it would be nice to have the option to see them in all their glory if only for posterity and to pick which is your favourite to work towards and maybe see a percentage progress towards unlocking all golds, all plat or poly? Just a thought 😄 

> > Can you think of a better way to do filtering without having 2 separate progress state/stores?
> 
> Not really sure what you mean here 🤔

Well with the design of the filtering and the shared code we removed in the review, I wanted your vue expertise on how we can best share methods like that filtering code between vue components that work with state.
Also with the requirements and default data stores I copied most of the layout from the way weapons are being stored and organised.
I feel like there could be an easier way if we restructured the way we're storing camo and weapon data so there's less code overall.
I'll have a further think about this and make a PR if I come up with something better.

**Comment by carlssonemil** [comment]
> Well with the design of the filtering and the shared code we removed in the review, I wanted your vue expertise on how we can best share methods like that filtering code between vue components that work with state. Also with the requirements and default data stores I copied most of the layout from the way weapons are being stored and organised. I feel like there could be an easier way if we restructured the way we're storing camo and weapon data so there's less code overall. I'll have a further think about this and make a PR if I come up with something better.

The easiest way would be to just add some new options to the filter variable in the store, make all options toggable through props in the filter component and then just for each view (weapon/camouflage) send in which filter options should be shown. The filtering itself is done in the View files, so it should be pretty easy. I have an idea in mind on how to solve all this, can look at it this weekend.

We should be careful with restructuring the weapon progress data though, since it'll be a pain to add some migrating code for users that have already input progress using the current implementation 🤔

**Review by carlssonemil — APPROVED** [review]
Looks good! Let's merge it! 🚀 

I'll look at the filter and favorite functionality this weekend! 😄

**Pull request merged**